File: go-apps/internal/modules/users/repository/user_repository.go
package repository

import (
	"context"
	"fmt"

	"go-apps/internal/modules/users/model"

	"gorm.io/gorm"
)

// UserRepository interface defines user repository operations
type UserRepository interface {
	Create(ctx context.Context, user *model.User) error
	GetByID(ctx context.Context, id uint) (*model.User, error)
	GetByEmail(ctx context.Context, email string) (*model.User, error)
	Update(ctx context.Context, user *model.User) error
	Delete(ctx context.Context, id uint) error
	GetAll(ctx context.Context, limit, offset int) ([]*model.User, error)
	GetActive(ctx context.Context) ([]*model.User, error)
	GetByRole(ctx context.Context, roleName string) ([]*model.User, error)
	Count(ctx context.Context) (int64, error)
	ExistsByEmail(ctx context.Context, email string) (bool, error)
	Search(ctx context.Context, query string, limit, offset int) ([]*model.User, int64, error)
}

// userRepository implements UserRepository interface
type userRepository struct {
	db *gorm.DB
}

// NewUserRepository creates a new user repository
func NewUserRepository(db *gorm.DB) UserRepository {
	return &userRepository{db: db}
}

// Create creates a new user
func (r *userRepository) Create(ctx context.Context, user *model.User) error {
	return r.db.WithContext(ctx).Create(user).Error
}

// GetByID gets a user by ID
func (r *userRepository) GetByID(ctx context.Context, id uint) (*model.User, error) {
	var user model.User
	err := r.db.WithContext(ctx).Preload("Roles").First(&user, id).Error
	if err != nil {
		if err == gorm.ErrRecordNotFound {
			return nil, fmt.Errorf("user not found")
		}
		return nil, err
	}
	return &user, nil
}

// GetByEmail gets a user by email
func (r *userRepository) GetByEmail(ctx context.Context, email string) (*model.User, error) {
	var user model.User
	err := r.db.WithContext(ctx).Preload("Roles").Where("email = ?", email).First(&user).Error
	if err != nil {
		if err == gorm.ErrRecordNotFound {
			return nil, fmt.Errorf("user not found")
		}
		return nil, err
	}
	return &user, nil
}

// Update updates a user
func (r *userRepository) Update(ctx context.Context, user *model.User) error {
	return r.db.WithContext(ctx).Save(user).Error
}

// Delete deletes a user
func (r *userRepository) Delete(ctx context.Context, id uint) error {
	return r.db.WithContext(ctx).Delete(&model.User{}, id).Error
}

// GetAll gets all users with pagination
func (r *userRepository) GetAll(ctx context.Context, limit, offset int) ([]*model.User, error) {
	var users []*model.User
	err := r.db.WithContext(ctx).Preload("Roles").
		Limit(limit).Offset(offset).
		Order("first_name ASC").
		Find(&users).Error
	return users, err
}

// GetActive gets all active users
func (r *userRepository) GetActive(ctx context.Context) ([]*model.User, error) {
	var users []*model.User
	err := r.db.WithContext(ctx).Preload("Roles").
		Where("is_active = ?", true).
		Order("first_name ASC").
		Find(&users).Error
	return users, err
}

// GetByRole gets users by role name
func (r *userRepository) GetByRole(ctx context.Context, roleName string) ([]*model.User, error) {
	var users []*model.User
	err := r.db.WithContext(ctx).Preload("Roles").
		Joins("JOIN user_roles ON users.id = user_roles.user_id").
		Joins("JOIN roles ON user_roles.role_id = roles.id").
		Where("roles.name = ?", roleName).
		Find(&users).Error
	return users, err
}

// Count gets the total count of users
func (r *userRepository) Count(ctx context.Context) (int64, error) {
	var count int64
	err := r.db.WithContext(ctx).Model(&model.User{}).Count(&count).Error
	return count, err
}

// ExistsByEmail checks if a user exists by email
func (r *userRepository) ExistsByEmail(ctx context.Context, email string) (bool, error) {
	var count int64
	err := r.db.WithContext(ctx).Model(&model.User{}).Where("email = ?", email).Count(&count).Error
	return count > 0, err
}

// Search searches users by query
func (r *userRepository) Search(ctx context.Context, query string, limit, offset int) ([]*model.User, int64, error) {
	var users []*model.User
	var total int64

	searchQuery := r.db.WithContext(ctx).Model(&model.User{}).
		Where("first_name ILIKE ? OR last_name ILIKE ? OR email ILIKE ?",
			"%"+query+"%", "%"+query+"%", "%"+query+"%")

	// Get total count
	if err := searchQuery.Count(&total).Error; err != nil {
		return nil, 0, err
	}

	// Get results
	err := searchQuery.Preload("Roles").
		Limit(limit).Offset(offset).
		Order("first_name ASC").
		Find(&users).Error

	return users, total, err
}

File: apps/api/src/modules/users/infrastructure/user.repository.ts
import { Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { User } from '../domain/entities/user.entity';
import { UserRepositoryInterface } from '../domain/repositories/user.repository.interface';

@Injectable()
export class UserRepository implements UserRepositoryInterface {
  constructor(
    @InjectRepository(User)
    private readonly repository: Repository<User>,
  ) {}

  async findAll(): Promise<User[]> {
    return this.repository.find({
      relations: ['roles'],
    });
  }

  async findById(id: number): Promise<User | null> {
    return this.repository.findOne({
      where: { id },
      relations: ['roles'],
    });
  }

  async findByEmail(email: string): Promise<User | null> {
    return this.repository.findOne({
      where: { email },
      relations: ['roles'],
    });
  }

  async create(user: Partial<User>): Promise<User> {
    const newUser = this.repository.create(user);
    return this.repository.save(newUser);
  }

  async update(id: number, user: Partial<User>): Promise<User> {
    await this.repository.update(id, user);
    return this.findById(id);
  }

  async delete(id: number): Promise<void> {
    await this.repository.delete(id);
  }

  async findActive(): Promise<User[]> {
    return this.repository.find({
      where: { isActive: true },
      relations: ['roles'],
      order: { firstName: 'ASC' },
    });
  }

  async findByRole(roleName: string): Promise<User[]> {
    return this.repository
      .createQueryBuilder('user')
      .leftJoinAndSelect('user.roles', 'roles')
      .where('roles.name = :roleName', { roleName })
      .getMany();
  }

  async findWithRoles(): Promise<User[]> {
    return this.repository.find({
      relations: ['roles'],
    });
  }

  async count(): Promise<number> {
    return this.repository.count();
  }

  async existsByEmail(email: string): Promise<boolean> {
    const count = await this.repository.count({ where: { email } });
    return count > 0;
  }

  async findPaginated(page: number, limit: number, search?: string): Promise<{ users: User[]; total: number }> {
    const queryBuilder = this.repository.createQueryBuilder('user')
      .leftJoinAndSelect('user.roles', 'roles');
    
    if (search) {
      queryBuilder.where(
        'user.firstName ILIKE :search OR user.lastName ILIKE :search OR user.email ILIKE :search',
        { search: `%${search}%` }
      );
    }
    
    const [users, total] = await queryBuilder
      .orderBy('user.firstName', 'ASC')
      .skip((page - 1) * limit)
      .take(limit)
      .getManyAndCount();
    
    return { users, total };
  }
}
